tests/e2e/utils.go

```diff
@@ -72,34 +81,64 @@ func NewFunnel() *Funnel {
 		},
 	}
 
-	conn, err := grpc.Dial(conf.RPCAddress(), grpc.WithInsecure())
-	if err != nil {
-		panic(err)
-	}
-
 	var derr error
 	dcli, derr := util.NewDockerClient()
 	if derr != nil {
 		panic(derr)
 	}
 
 	return &Funnel{
-		RPC:        tes.NewTaskServiceClient(conn),
 		HTTP:       client.NewClient("http://localhost:" + conf.HTTPPort),
 		Docker:     dcli,
 		Conf:       conf,
 		StorageDir: storageDir,
 		startTime:  fmt.Sprintf("%d", time.Now().Unix()),
 		rate:       rate,
-		conn:       conn,
 	}
 }
 
-// Tempdir returns a new temporary directory path
-func (f *Funnel) Tempdir() string {
-	d, _ := ioutil.TempDir(f.StorageDir, "")
-	d, _ = filepath.Abs(d)
-	return d
+func (f *Funnel) addRPCClient() {
+	conn, err := grpc.Dial(f.Conf.RPCAddress(), grpc.WithInsecure())
+	if err != nil {
+		panic(err)
+	}
+	f.conn = conn
+	f.RPC = tes.NewTaskServiceClient(conn)
+}
+
+// StartServer starts the server
+func (f *Funnel) StartServer() {
+	go server.Run(f.Conf)
+	f.pollForServerStart()
+	f.addRPCClient()
+}
+
+func (f *Funnel) pollForServerStart() {
+	timeout := make(chan bool, 1)
+	go func() {
+		time.Sleep(30 * time.Second)
+		timeout <- true
+	}()
+
+	srvRunning := make(chan bool, 1)
+	go func() {
+		for {
+			_, err := f.HTTP.GetServiceInfo()
```

Comment:
Or even, use the RPC client and set it to retry, add a context with a timeout, and possibly you can collapse all this into a single line.